test/Dialect/FIRRTL/infer-widths-errors.mlir

```diff
@@ -121,3 +121,16 @@ firrtl.circuit "Foo"  {
     firrtl.connect %w, %0 : !firrtl.bundle<x: uint>, !firrtl.bundle<x: uint>
   }
 }
+
+// -----
+// Unsatisfiable widths through AttachOp on analog types should error.
+// https://github.com/llvm/circt/issues/4786
+firrtl.circuit "AnalogWidths" {
+  // expected-error @+1 {{uninferred width: port "a" cannot satisfy all width requirements}}
+  firrtl.module @AnalogWidths(in %a: !firrtl.analog, out %b: !firrtl.analog<2>, out %c: !firrtl.analog<1>) {
+    firrtl.attach %a, %b : !firrtl.analog, !firrtl.analog<2>
+    // expected-note @+2 {{width is constrained to be at least 2 here:}}
+    // expected-note @+1 {{width is constrained to be at most 1 here:}}
```

Comment:
Oh, sorry.  Don't move the expected comments-- they still need to be pointing at the 'attach'es.  "@below" just lets you stack them without counting and seems a little more readable than +/-1 ("@above").

Out of curiosity, why are both note diagnostics located on the second attach ("at most 1" seems from the first attach)? Not a big deal just curious.  Anyway if using `@below` both notes would go above that.  For convenience since touched this up locally while poking at it:

Reply:
Haha I'm such an idiot. Thanks for clarifying :100:, and thanks for the fix :smile: !

Reply:
> Out of curiosity, why are both note diagnostics located on the second attach ("at most 1" seems from the first attach)?

That stems from a bit of a design smell in `InferWidths` at the moment: the expressions being created in the solver are deduplicated, which makes it hard to pinpoint the *exact* location of where a constraint comes from. There's a bit of hackery going on where the pass roughly keeps track of the locations associated with a constraint, but that doesn't always point to the interesting thing if two expressions deduplicate (which is quite common for constants like `2` in the above case).

Ideally we wouldn't have to build up an entire expression graph to do the width inference, doing it on the ops directly instead. But I haven't had a chance to look into that yet :cry:. We could also just stop deduplicating expressions to be more precise in diagnostics, but that likely also increases the memory footprint of the pass (because it roughly clones the FIRRTL IR into this expression graph).